### ISSUE TITLE
INTERNAL: Protocol error Rst Stream when communicating with .NET Core 3 server.

### ISSUE THREAD

**Issue by ivan-penchev** [grpc/grpc-java#6202]
I have compiled a repo with the issue in a reproducible manner : 
https://github.com/ivan-penchev/gRPC-Java-dotnetcore3/tree/master

I have basically download the "Greeter" example from dotnet and java  repos and tried to make them work together. Since we are trying to migrate our services to .NET Core 3.0;

Someone reported the same issue [on stackoverflow for Kotlin](https://stackoverflow.com/questions/57414584/rst-stream-grpc-error-on-android-kotlin-client-when-requesting-net-core-3-0-grp)

I have tried the same .NET Core 3 server with Python, .NET and BloomRPC client and it worked flawlessly, which is why I post the issue here.
### What version of gRPC are you using?
Java uses : 1.23.0
.NET uses: 2.23.2

### What operating system (Linux, Windows,...) and version?
I am running on Windows, using Docker in a Linux container.

### What did you expect to see?
I expected to see a response containing "Hello ~whatever message I sent initially~".

### What did you see instead?

```java
Failed... : 
io.grpc.StatusRuntimeException: INTERNAL: Protocol error
Rst Stream
	at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:235)
	at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:216)
	at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:141)
	at io.grpc.examples.helloworld.GreeterGrpc$GreeterBlockingStub.sayHello(GreeterGrpc.java:228)
	at io.grpc.helloworldexample.HelloworldActivity$GrpcTask.doInBackground(HelloworldActivity.java:90)
	at io.grpc.helloworldexample.HelloworldActivity$GrpcTask.doInBackground(HelloworldActivity.java:72)
	at android.os.AsyncTask$2.call(AsyncTask.java:333)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:764)

```

### Steps to reproduce the bug

I have described it completely [here](https://github.com/ivan-penchev/gRPC-Java-dotnetcore3/blob/master/README.md).  

I have also attached a log from the package sniffer on the client [here](https://github.com/ivan-penchev/gRPC-Java-dotnetcore3/blob/master/client-packet-sniff-full).

**Comment by ivan-penchev** [comment]
Hi @creamsoup, 
Just wanted to check if you manage to reproduce the problem using the provided repo?

Also a quck note: I have updated the repo with the "possible" solution provided from the .net grpc team, and the issue still persists.

Thanks.

**Comment by JamesNK** [comment]
I have recreated the issue, and the problem is the Java client is sending an `https` scheme when making h2c (plaintext) gRPC calls:

> Trace id "0HLQ948TUS99R:00000003": HTTP/2 stream error "PROTOCOL_ERROR". A Reset is being sent to the stream.
Microsoft.AspNetCore.Connections.ConnectionAbortedException: The request :scheme header 'https' does not match the transport scheme 'http'.

The .NET Core 3 server is validating that the request scheme matches the transport and is rejecting the request.

![image](https://user-images.githubusercontent.com/303201/66245916-554be700-e76d-11e9-9c9e-6a3b3b3ecba1.png)

**Comment by sanjaypujare** [comment]
Looking at [OkHttpClientTransport.createHttpProxyRequest](https://github.com/grpc/grpc-java/blob/master/okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java#L690) I see "https" hardcoded 

```
    HttpUrl tunnelUrl = new HttpUrl.Builder()
        .scheme("https")
        .host(address.getHostName())
        .port(address.getPort())
        .build();
```
regardless of your usePlaintext() on the channel-builder. Not sure if you are using a proxy (unknowingly). But I see the destination address (in wireshark) and authority are different. So may be?

**Comment by JamesNK** [comment]
I don't know. I'm not familiar with how the Java gRPC client works. The original comment links to a source code repro.

Searching around:

There is this which is also hardcoded: https://github.com/grpc/grpc-java/blob/b0f423295b4674cb5247a6143fd211b050ef0065/okhttp/src/main/java/io/grpc/okhttp/Headers.java#L37

And this one, although it is related to TLS negotiation so hardcoding to https probably makes sense: https://github.com/grpc/grpc-java/blob/c9177b2327c8c424af7889d2306a4c767455f10b/alts/src/main/java/io/grpc/alts/internal/AltsProtocolNegotiator.java#L61

**Comment by sanjaypujare** [comment]
The AltsProtocolNegotiator doesn't matter. But the one in [Headers.java ](https://github.com/grpc/grpc-java/blob/b0f423295b4674cb5247a6143fd211b050ef0065/okhttp/src/main/java/io/grpc/okhttp/Headers.java#L37)likely seems to be the cause.

**Comment by ivan-penchev** [comment]
@sanjaypujare  there is no proxy involved (at least to my knowledge).
I run the android app on the phone, and connect to my PC locally via Wi-Fi. The PC is Windows runing a docker image of the .NET core server.

**Comment by sanjaypujare** [comment]
@ivan-penchev it might not be proxy related. Could be the hardcoding in [Headers.java ](https://github.com/grpc/grpc-java/blob/b0f423295b4674cb5247a6143fd211b050ef0065/okhttp/src/main/java/io/grpc/okhttp/Headers.java#L37) that @JamesNK pointed out and may need to be fixed

**Comment by JamesNK** [comment]
Can this be fixed in 1.25? A couple of people have run into this issue, and it isn't obvious from the error for the user to know how to fix it.

**Comment by ran-su** [comment]
@JamesNK Yes, we will fix the okhttp issue before 1.25 release and back-port the fix. However, we still think enforcing the scheme match might be the thing we do not want to do. I will create an issue about this on c#'s repo and discuss our view of it.

**Comment by JamesNK** [comment]
Great!

Validation happens inside the .NET web server. Add your thoughts about it here - https://github.com/aspnet/AspNetCore/issues/14745
